### PR TITLE
Do not add an unused conditional field in packet fields

### LINKED ISSUE #2363: Error when dissecting packet with conditional fields
#### Brief description

When dissecting a packet, fields that don't fulfill a condition can have the value None.

#### Environment

-   Scapy version: last commit 236102447f41f9818a09cde91c83af472b9b7383 but also old scapy such as version 2.3.3
-   Python version: python2 and python3
-   Operating System: independent of th OS

#### How to reproduce

```
class TEST_COND(Packet):
    fields_desc = [
        IntField("A", 0),
        ConditionalField(IntField("A0",0), lambda pkt:pkt.A == 0),
        ConditionalField(IntField("A1",0), lambda pkt:pkt.A != 0),
        IntField("B", 0),
        ConditionalField(IntField("B0",0), lambda pkt:pkt.B == 0),
        ConditionalField(IntField("B1",0), lambda pkt:pkt.B != 0),
        ]

TEST_COND(TEST_COND().build()).fields
```

#### Actual result

```
{'A': 0, 'A0': 0, 'B': 0, 'A1': None, 'B0': 0}
```

#### Expected result

```
{'A': 0, 'A0': 0, 'B': 0, 'B0': 0}
```

#### Related resources

None

### PR DESCRIPTION
**Pull request by guedou**
This PR fixes #2363.

### CODE REVIEW

**Comment by LRGH** [comment]
Thanks. I see that you chose my first proposal, of not including the fields for which the condition is not True, and therefore the non-regression tests needed to be changed.

But I think that your patch is not perfect: it will create problems in case there is a conditional field, with True condition, and None value.
I would prefer
```
--- a/scapy/packet.py
+++ b/scapy/packet.py
@@ -831,6 +831,8 @@ class Packet(six.with_metaclass(Packet_metaclass, BasePacket,
         for f in self.fields_desc:
             if not s:
                 break
+            if isinstance(f, ConditionalField) and not f._evalcond(self):
+                continue
             s, fval = f.getfield(self, s)
             self._tmp_dissect_pos = _lr - len(s)
             # We need to track fields with mutable values to discard
```
Note that the test `if isinstance(f, ConditionalField) and not f._evalcond(self):` appears in many other places in packet.py, it seems the standard way to do this test.

**Comment by guedou** [comment]
During dissection, `fval` cannot be `None`. That's why I decided to skip the test.

**Comment by LRGH** [comment]
It might be a requirement of scapy, but it is not always true. In the example below, fval is None when parsing field "qd". Note that I agree that it can be seen as a bug in layers/dns.py 
```
from scapy.all import *
class DNSRR(Packet):
    fields_desc = [
        DNSRRCountField("qdcount", None, "qd"),
        ConditionalField(DNSRRField("qd", "qdcount"), lambda pkt:True)
    ]

DNSRR("\0\0d")
```

**Comment by guedou** [comment]
This seems fine when dissecting. I don't understand the issue here.

**Comment by LRGH** [comment]
The example with DNSRR is not an example where dissection fails, it is an example where the assumption "During dissection, `fval` cannot be `None`" is not met.

With the current, unpatched scapy, we have
```DNSRR("\0\0d").fields == {'qdcount': 0, 'qd': None}```
With your patch we have
```DNSRR("\0\0d").fields == {'qdcount': 0}```

This does not cause any issue, because the default value of a `DNSRRField` is `None`, and if there is no value in `packet.fields` then the default value is used.

But your patch would break the use of a `ConditionalField` with a field such that dissecting can return `None` and the default value is not `None`. I did not find such an example is the standard scapy distribution, but some user of scapy may have created a field with these properties.

**Comment by LRGH** [comment]
The issue with conditional fields is even worse than I thought.
I have updated the description of my original issue.

**Comment by gpotter2** [comment]
Sorry I did not meant to push on your repo. I was going to open a new PR.
I guess we'll keep this one for now. What do you think @guedou

**Comment by guedou** [comment]
@gpotter2 this looks good to me. Do you want to take over the PR? I won't have time to investigate soon.

**Review by p-l- — DISMISSED** [review]

**Comment by gpotter2** [comment]
Finally got back to this. There still was an issue when combining a `CondionalField` and a `MultipleTypeField`. Everything should be working properly now

**Comment by gpotter2** [comment]
@p-l- @guedou Hi, could you please review this? Thanks

File: scapy/contrib/macsec.py [hunk 1]
```diff
@@ -210,19 +213,35 @@ def decrypt(self, orig_pkt, assoclen=None):
...
-                   BitField('SCB', 0, 1),
-                   BitField('E', 0, 1),
-                   BitField('C', 0, 1),
-                   BitField('an', 0, 2),
-                   BitField('reserved', 0, 2),
-                   BitField('shortlen', 0, 6),
-                   IntField("pn", 1),
-                   ConditionalField(PacketField("sci", None, MACsecSCI), lambda pkt: pkt.SC),  # noqa: E501
-                   ConditionalField(XShortEnumField("type", None, ETHER_TYPES),
-                                    lambda pkt: pkt.type is not None)]
```
Comment (gpotter2) [inline scapy/contrib/macsec.py:225]: The issue here was this lambda referencing itself...

File: scapy/fields.py [hunk 2]
```diff
@@ -309,6 +309,27 @@ def _evalcond(self, pkt):
...
+        # BACKWARD COMPATIBILITY
+        # Note: we shouldn't need this function. (it's not correct)
+        # However, having i2h implemented (#2364), it changes the default
+        # behavior and broke all packets that wrongly use two ConditionalField
+        # with the same name. Those packets are the problem: they are wrongly
+        # built (they should either be re-using the same conditional field, or
+        # using a MultipleTypeField).
+        # But I don't want to dive into fixing all of them just yet,
+        # so for now, let's keep this this way, even though it's not correct.
+        if type(self.fld) == Field:
```
Comment (p-l-) [inline scapy/fields.py:323]:
```suggestion
        if isinstance(self.fld, Field):
```

**Review by guedou — COMMENTED** [review]
I agree with @p-l- changes requests.

File: scapy/fields.py (hunk 2 above)
Comment (gpotter2) [reply scapy/fields.py]: The latter. Thanks

File: scapy/fields.py [hunk 3]
```diff
@@ -309,6 +309,27 @@ def _evalcond(self, pkt):
...
+        # BACKWARD COMPATIBILITY
+        # Note: we shouldn't need this function. (it's not correct)
+        # However, having i2h implemented (#2364), it changes the default
+        # behavior and broke all packets that wrongly use two ConditionalField
+        # with the same name. Those packets are the problem: they are wrongly
+        # built (they should either be re-using the same conditional field, or
+        # using a MultipleTypeField).
+        # But I don't want to dive into fixing all of them just yet,
+        # so for now, let's keep this this way, even though it's not correct.
+        if type(self.fld) is Field:
```
Comment (guedou) [inline scapy/fields.py:323]: Codacity complains about this line too =)

Comment (gpotter2) [reply scapy/fields.py:323]: Except they're wrong. I only want Field to be detected, not any of its subfields.

**Review by p-l- — APPROVED** [review]

**Pull request merged**